### PR TITLE
[T5] lm_head weights initialization: set variance to reciprocal of hidden dim

### PR DESCRIPTION
**Pull request by Birch-san**
# What does this PR do?

image

before this PR: lm_head weights were initialized with variance of 1, and it output activations with variance ~= hidden_dim. this is a very high variance for logits, and resulted in initial cross-entropy loss of ~110, which is Very High.

after this PR: lm_head weights initialized with variance of reciprocal of hidden_dim. this outputs activations with variance ~= 1. this is results in initial cross-entropy loss of ~11, which is high, but in line with what we'd expect.

Fixes https://github.com/huggingface/transformers/issues/16749 (again)

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [ ] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md#start-contributing-pull-requests),
      Pull Request section?
- [ ] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [ ] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/main/docs#writing-source-documentation).
- [ ] Did you write any new necessary tests?

## Who can review?

@ArthurZucker @younesbelkada

### CODE REVIEW

**Review by younesbelkada — COMMENTED** [review]
Thanks a lot for the deep investigation and making T5 models converge better ! Can you run the styling checks?
```bash
make fixup && make fix-copies
```
Then we can merge I think

**Comment by abdulfatir** [comment]
I would like to argue for setting the initialization of both input and output embedding layers to `normal(0, 0.05)` as discussed in this comment: https://github.com/huggingface/transformers/issues/16749#issuecomment-1101607487. I am not sure why this was actually not done (cc @patrickvonplaten).

**Comment by abdulfatir** [comment]
To provide more context: neither the default initialization (`normal(0, 1)`), not what's proposed in this PR works in my when experiment when I train from random weights. `normal(0, 0.05)` works and should probably be the default to align with the original implementation.

**Comment by Birch-san** [comment]
I wasn't sure what to make of Patrick's `stddev=0.05`, as no explanation was given.

I surmise like the reasoning was:

- `VocabEmbedding#__init__` initializes embedding [via the `mtf.layers.embedding_weights` helper](https://github.com/tensorflow/mesh/blob/a32810e32709e0eaad3b241475d3be0957409adc/mesh_tensorflow/transformer/transformer.py#L2155).  
  It **might** default to `_scale_variable_like_classifier_weights=False` (I didn't see any `.gin` file overriding this, but then again I never found the `.gin` file used to train T5), which would make `initializer=None`.
- when `initializer is None`, `mtf.layers.embedding_weights` initializes the embedding [via  `tf.random_normal_initializer()`](https://github.com/tensorflow/mesh/blob/a32810e32709e0eaad3b241475d3be0957409adc/mesh_tensorflow/layers.py#L2096).
- `tf.random_normal_initializer()` [defaults to std=0.05](https://www.tensorflow.org/api_docs/python/tf/random_normal_initializer).

That said, when `hidden_dim==512` (as in t5-small): std=0.05 is _awfully_ close to `std=hidden_dim**-.5=0.044`.

I tried initializing both the embedding and the lm_head with std=0.05. It performed moreorless the same (well, fractionally worse).

image

**Comment by abdulfatir** [comment]
Thanks for the quick turnaround. In my experiment:
- Current default `std=1`: Starts at a high loss and converges to high value.
- `lm_head` with `std=hidden_dim**-.5`: Starts at a lower loss than before but diverges after a few hundred steps.
- `initializer_factor=0.05`: Starts at an even lower loss and converges nicely.

**Comment by Birch-san** [comment]
how big is your hidden_dim? for t5-small, `hidden_dim**-.5=0.044` is a very similar number to 0.05.

**Comment by abdulfatir** [comment]
I am also running my quick tests on `flan-t5-small`.

**Comment by Birch-san** [comment]
okay yeah certainly I agree that we should modify _both_ the embedding **and** the lm_head weights.

I think we don't _actually_ have any info from the mesh tensorflow repository on how to initialize an _untied_ `lm_head`. I don't think they support that.

**Comment by Birch-san** [comment]
I tried a model with a bigger hidden dim (t5 base, 768). so that there would be a bigger difference between 0.05 and `hidden_dim**-.5`.

image

no discernable difference by 600 steps.

so I don't know which is more _correct_ out of `dim**-.5` or `0.05`, but I think empirically there's not a big difference.

certainly I think initializing the _embedding_ to 0.05 matches the MTF repository.

but I don't think the MTF repository supports an untied head at all, so I think that's up to us? unless there's some precedent for what to do there. there's two schools of thought:

- initialize the lm_head as though it were tied to the embedding (0.05)
- initialize the lm_head like you'd initialize most Linear layers (`dim**-.5`)

I kinda think treating the lm_head as untied (`dim**-.5`) makes more sense.

numerically and empirically I think there's not much difference. I think the more important thing is don't initialize either the embedding or the lm_head with `std=1`.

**Comment by Birch-san** [comment]
oh, there _is_ a reference to tied vs untied lm_head:  
https://github.com/tensorflow/mesh/blob/fa19d69eafc9a482aff0b59ddd96b025c0cb207d/mesh_tensorflow/transformer/transformer.py#L586-L594

I think they're saying "if the embedding is tied, we anticipate that the hidden states will have too much variance, and so we correct it". so whatever std they use to initialize the embedding, is higher than the std an lm_head would want?

**Comment by Birch-san** [comment]
maybe we can learn something about how to initialize the weights, based on how they ended up distributed after Google finished training the model?

I took an off-the-shelf model:

```python
from transformers import T5ForConditionalGeneration
model: T5ForConditionalGeneration = T5ForConditionalGeneration.from_pretrained('google/t5-v1_1-small')
```

and I [analysed how all of its layers were distributed](https://gist.github.com/Birch-san/7709286ec73c6795666050a4f2786309).

embedding std is 11.6375  
lm_head std is 1.1614

considering encoder self-attn, decoder self-attn, decoder cross-attn…  
attn q is ~0.05 or ~`hidden_dim**-.5` (hard to tell the difference)  
attn k is 0.3~0.4  
attn v is 0.6~0.8  
attn o is 0.6~1.0

considering encoder FFN, decoder FFN…  
FFN wi_0 is 0.32~0.36  
FFN wi_1 is 0.85~1.17  
FFN wo is 0.53~0.73

encoder per-block layernorms ~0.02  
encoder final layernorm ~0.05 or ~`hidden_dim**-.5` (hard to tell the difference)  

decoder per-block layernorms 0.07~0.10  
decoder final layernorm ~0.7

**Comment by Birch-san** [comment]
t5-small pretraining `.gin` config apparently [lives here](https://storage.cloud.google.com/t5-data/pretrained_models/small/operative_config.gin?_ga=2.225153114.-1337093425.1644544368).  
_from https://github.com/allenai/unifiedqa/issues/40#issuecomment-1049151987._

File: src/transformers/models/t5/modeling_t5.py [hunk 1]
```diff
@@ -829,7 +829,7 @@ def _init_weights(self, module):
             # See https://github.com/tensorflow/mesh/blob/fa19d69eafc9a482aff0b59ddd96b025c0cb207d/mesh_tensorflow/layers.py#L1624
             module.shared.weight.data.normal_(mean=0.0, std=factor * 1.0)
             if hasattr(module, "lm_head") and not self.config.tie_word_embeddings:
-                module.lm_head.weight.data.normal_(mean=0.0, std=factor * 1.0)
+                module.lm_head.weight.data.normal_(mean=0.0, std=factor * self.config.d_model ** -.5)
```
Comment (patrickvonplaten) [inline src/transformers/models/t5/modeling_t5.py:832]:
```suggestion
                module.lm_head.weight.data.normal_(mean=0.0, std=factor * 0.05)
```

To align with the origin implementation, let's multiply all lm head inits with 0.05, not sure why this wasn't done back then. 

@Birch-san could you maybe change this here: https://github.com/huggingface/transformers/blob/63864e057fd4ecbf54c77599702873f7be871e65/src/transformers/models/t5/modeling_flax_t5.py#L1512 and here: https://github.com/huggingface/transformers/blob/1b8decb04c246ec8e1c4ba7f2749043d0876d24e/src/transformers/models/t5/modeling_tf_t5.py#L1249 as well?

**Comment by patrickvonplaten** [comment]
The T5X repository has become somewhat of a reference implementation to pretrain T5X. Here they use stddev=1.0 as an initialization: https://github.com/google-research/t5x/blob/b051e46075fdcb02fcdd4dc648dd9560243bfdb2/t5x/examples/t5/network.py#L289

That aligns with what we currently have in Transformers. Should we maybe open an issue in T5X asking about the differences because it indeed seems like in the original implementation they used 0.05 as init.

**Comment by Birch-san** [comment]
@patrickvonplaten  
> Here they use stddev=1.0 as an initialization

do they though? it looks to me like they use `scale=1.0`, not `std=1.0`.

- since embedding is [untied](https://github.com/google-research/t5x/blob/b051e46075fdcb02fcdd4dc648dd9560243bfdb2/t5x/examples/t5/t5_1_1/base.gin#L55),
- [lm_head is a DenseGeneral](https://github.com/google-research/t5x/blob/b051e46075fdcb02fcdd4dc648dd9560243bfdb2/t5x/examples/t5/network.py#L269) with default initialization
- DenseGeneral's default initialization [is `variance_scaling(1.0, 'fan_in', 'truncated_normal')`](https://github.com/google-research/t5x/blob/b051e46075fdcb02fcdd4dc648dd9560243bfdb2/t5x/examples/t5/layers.py#L351)
- `1.0` [is the `scale` parameter](https://jax.readthedocs.io/en/latest/_autosummary/jax.nn.initializers.variance_scaling.html) in `std=(scale/n)**.5`
  - when `mode='fan_in'`, `n` refers to "the number of input units in the weights tensor", which I guess is equivalent to PyTorch's `Linear(in_features=…)`
  - so `n=hidden_dim`
  - in other words: `std=hidden_dim**-.5`

**Comment by Birch-san** [comment]
@patrickvonplaten  
> Should we maybe open an issue in T5X asking about the differences because it indeed seems like in the original implementation they used 0.05 as init.

what did you think of my investigation that suggested that MTF [initialized lm_head using `std=hidden_dim**-.5`](https://github.com/huggingface/transformers/pull/26441#issuecomment-1741877125)?

**Comment by patrickvonplaten** [comment]
> The T5X repository has become somewhat of a reference implementation to pretrain T5X. Here they use stddev=1.0 as an initialization: https://github.com/google-research/t5x/blob/b051e46075fdcb02fcdd4dc648dd9560243bfdb2/t5x/examples/t5/network.py#L289
> 
> That aligns with what we currently have in Transformers. Should we maybe open an issue in T5X asking about the differences because it indeed seems like in the original implementation they used 0.05 as init.

Agree that there is a difference! Let's maybe open an issue in T5X and link it here? Think T5X is still pretty active

**Comment by Birch-san** [comment]
https://github.com/google-research/t5x/issues/1477

**Comment by patrickvonplaten** [comment]
Thanks a lot for posting on T5X @Birch-san - :crossed_fingers: someone from the T5 team looks into it

**Comment by vadimkantorov** [comment]
maybe worth un-staling/re-opening this issue?

**Comment by vadimkantorov** [comment]
Haha it appears the stale bot is over-aggressive

**Comment by dhruvbird** [comment]
I ran into exactly this issue today, and the proposed fixed made the grad_norm during training misbehave less (at least so far. Here are the charts for reference. 

image

Purple is on master, and blue is with the fix in this MR. 

It would be super if we can merge this PR into master so that anyone trying to train a T5 model from scratch (i.e. no fine-tuning using pre-trained weights) can benefit from this fix.

**Comment by amyeroberts** [comment]
@dhruvbird Thanks for sharing your results! It appears this wasn't merged in as it was pending confirmation from the T5 team. Although this would technically be a breaking change, it does appear to be useful for training the model. WDYT about adding @ArthurZucker ?

**Comment by ArthurZucker** [comment]
Yeah sounds good in general, the issue is that it could affect the entire field of diffusion that relies on t5 cc @sayakpaul

**Comment by sayakpaul** [comment]
In the diffusion community, T5 is not generally trained from scratch, though. Just a heads up.

**Comment by Birch-san** [comment]
I also note that attention is not being scaled by the usual `head_dim**-.5`:  
https://github.com/huggingface/transformers/blob/1bd9d1c89998d7cf7f1ce0290e8944e9299cf407/src/transformers/models/t5/modeling_t5.py#L525

instead, this scale factor is fused into the weight init of the Q projection:  
https://github.com/huggingface/transformers/blob/1bd9d1c89998d7cf7f1ce0290e8944e9299cf407/src/transformers/models/t5/modeling_t5.py#L864

it's likely Google did this for performance reasons (I expect only marginal, and certainly outmoded by flash attention).  
they made this decision knowing they were using Adafactor optimizer, which exhibits equivalent training dynamics either way.

under AdamW, this optimization is harmful. you can only do this trick if your optimizer scales lr by the RMS of the params (e.g. Adafactor, [LARS+AdamW](https://github.com/kakaobrain/torchlars/blob/3b3d7e9c7bd35a31b2c2fa6f213beb0bf6881892/torchlars/lars.py#L133), or [AdamWScaled](https://github.com/PiotrNawrot/nanoT5/blob/1c82d67bf8dea635be68a3b2a68a43b68b665193/nanoT5/utils/copied_utils.py#L339C17-L339C26)).

so I also recommend to change the Q weight init to be the same as the K and V weight init, and to scale `Q•K.T` by `head_dim**-.5` (i.e. typical attention formulation). upgrading to torch SDPA would of course be nice too.

**Comment by ArthurZucker** [comment]
#31167 is coming, I gotta review it, scaling would be breaking as it is not done currently
